src/main/scala/pricemigrationengine/services/CohortTable.scala

```diff
@@ -1,19 +1,69 @@
 package pricemigrationengine.services
 
+import java.util
+
+import com.amazonaws.services.dynamodbv2.model.{AttributeValue, QueryRequest}
+import pricemigrationengine.dynamodb.{DynamoDBZIO, DynamoDBZIOError}
 import pricemigrationengine.model._
-import zio.ZIO
+import zio.stream.ZStream
+import zio.{IO, UIO, ZIO, ZLayer}
+
+import scala.jdk.CollectionConverters._
+
 
 object CohortTable {
 
   trait Service {
-    def fetch(filter: CohortTableFilter, batchSize: Int): ZIO[Any, CohortFetchFailure, Set[CohortItem]]
+    def fetch(
+      filter: CohortTableFilter,
+      batchSize: Int
+    ): IO[CohortFetchFailure, ZStream[Any, CohortFetchFailure, CohortItem]]
 
     def update(result: EstimationResult): ZIO[Any, CohortUpdateFailure, Unit]
   }
 
-  def fetch(filter: CohortTableFilter, batchSize: Int): ZIO[CohortTable, CohortFetchFailure, Set[CohortItem]] =
+  def fetch(filter: CohortTableFilter, batchSize: Int): ZIO[CohortTable, CohortFetchFailure, ZStream[Any, CohortFetchFailure, CohortItem]] =
     ZIO.accessM(_.get.fetch(filter, batchSize))
 
   def update(result: EstimationResult): ZIO[CohortTable, CohortUpdateFailure, Unit] =
     ZIO.accessM(_.get.update(result))
+
+  val impl: ZLayer[DynamoDBZIO, Nothing, UserRepo] =
+    ZLayer.fromFunction { dynamoDBZStream: DynamoDBZIO =>
```

Comment:
Im interested in peoples opinions of this pattern. 

It seems if you use the layer 'Acccessor' methods eg [here](https://github.com/guardian/price-migration-engine/blob/683dd57b47802124dbe50d69dd41d6b38e9281fa/src/main/scala/pricemigrationengine/services/Logging.scala#L13) in the service methods eg in this case 'fetch' it forces you to include the Logging service as an input for the IO returned by the service method.

Eg the fetch interface becomes:
```
def fetch(... ): ZIO[DynamoDBZIO, CohortFetchFailure, ZStream[Any, CohortFetchFailure, CohortItem]]
```

Which is exposing the implementation in the interface and there would be a different interface for different implementations.

Its weird as it seems to make the accessors only useful at the top level.

Am I missing something, is there a better way of doing this.

Reply:
I think the point is to distinguish between the environment required to build a service and the environment required to use the service.  So, for example, a live implementation would generally need a configuration service to build it but the interface wouldn't need to know about configuration as it's not needed to make use of the service.
So Logging should probably be part of the interface spec unless the service only logs while it's being built.

Reply:
Hum... For example the ConsoleLogging service is dependent on the console which it does nothing with at creation time, only using it when you 'use' a logging method.

You wouldnt add it to the interface as you would need to supply a console dependency to the Lambda logging.

And looking at the implementation you have done the same thing as I did here, ie you dont use the Console 'accessor' methods as that seems to force you to add the dependencies to the interface.

Thinking about you could use the accessors and then provide the dependencies eg:
```scala
object ConsoleLogging {
  val impl: ZLayer[Console, Nothing, Logging] = ZLayer.fromFunction(
    consoleService =>
      new Logging.Service {
        def info(s: String): UIO[Unit] = console.putStrLn(s"INFO: $s").provide(consoleService)
        def error(s: String): UIO[Unit] = console.putStrLn(s"ERROR: $s").provide(consoleService)
    }
  )
}
``` 

But that maybe seems a bit less readable, not sure

Reply:
Yes I was thinking about it again, and from a method that has the service as part of its environment it might become a bit unwieldy if you have to include the dependencies of the services you're calling as well as the services you're making use of directly.  So you probably wouldn't expect the return type of a method to be `ZIO[... with Logging, ..., ...]` unless you were actually using `Logging` in your method.

So this is fine as it is. 👍